core/embed/rust/src/ui/model_tt/component/hs_render.rs

```diff
@@ -0,0 +1,578 @@
+            4,
+            text_info.0.x0,
+            text_info.1,
+        );
+    }
+
+    y_tmp == (text_info.0.y1 - 1)
+}
+
+fn homescreen_get_fg_icon(
+    y_tmp: i16,
+    text_info: HomescreenTextInfo,
+    icon_data: &[u8],
+    fg_buffer: &mut LineBuffer4Bpp,
+) {
+    if let Some(icon_area) = text_info.3 {
+        let icon_size = icon_area.size();
+        if y_tmp >= icon_area.y0 && y_tmp < icon_area.y1 {
+            let y_pos = y_tmp - icon_area.y0;
+            position_buffer(
+                &mut fg_buffer.buffer,
+                &icon_data
+                    [(y_pos * icon_size.x / 2) as usize..((y_pos + 1) * icon_size.x / 2) as usize],
+                4,
+                icon_area.x0,
+                icon_size.x,
+            );
+        }
+    }
+}
+
+fn homescreen_position_text(
+    text: HomescreenText,
+    buffer: &mut BufferText,
+    icon_buffer: &mut [u8],
+) -> HomescreenTextInfo {
+    let (text, text_style, text_offset, icon) = text;
+
+    let text_width = display::text_width(text, text_style.text_font.into());
+    let font_max_height = display::text_max_height(text_style.text_font.into());
+    let font_baseline = display::text_baseline(text_style.text_font.into());
+    let text_width_clamped = text_width.clamp(0, screen().width());
+
+    let icon_size = if let Some(icon) = icon {
+        let (icon_size, icon_data) = toif_info_ensure(icon, ToifFormat::GrayScaleEH);
+        assert!(icon_size.x < HOMESCREEN_MAX_ICON_SIZE);
+        assert!(icon_size.y < HOMESCREEN_MAX_ICON_SIZE);
+        let mut ctx = UzlibContext::new(icon_data, None);
+        unwrap!(ctx.uncompress(icon_buffer), "Decompression failed");
+        icon_size
+    } else {
+        Offset::zero()
+    };
+
+    let text_top = screen().y0 + text_offset.y - font_max_height + font_baseline;
+    let text_bottom = screen().y0 + text_offset.y + font_baseline;
+    let icon_left = screen().center().x - (text_width_clamped + icon_size.x + TEXT_ICON_SPACE) / 2;
+    let text_left = icon_left + icon_size.x + TEXT_ICON_SPACE;
+    let text_right = screen().center().x + (text_width_clamped + icon_size.x + TEXT_ICON_SPACE) / 2;
+
+    let text_area = Rect::new(
+        Point::new(text_left, text_top),
+        Point::new(text_right, text_bottom),
+    );
+
+    let icon_area = if icon.is_some() {
+        Some(Rect::from_top_left_and_size(
+            Point::new(icon_left, text_bottom - icon_size.y - font_baseline),
+            icon_size,
+        ))
+    } else {
+        None
+    };
+
+    display::text_into_buffer(text, text_style.text_font.into(), buffer, 0);
+
+    (text_area, text_width, text_style.text_color, icon_area)
+}
+
+fn homescreen_line_blurred(
+    icon_data: &mut [u8],
+    text_buffer: &mut BufferText,
+    text_info: HomescreenTextInfo,
+    totals: [[u16; HOMESCREEN_IMAGE_SIZE as usize]; COLORS],
+    y: i16,
+) -> bool {
+    let t_buffer = unsafe { get_buffer_4bpp((y & 0x1) as u16, true) };
```

Comment:
hm the buffer safety is now left to caller since its where the responsibility not to request a buffer twice is. 

thinking we could do some tracking which buffers are already used (track it in the C part), and return null when requesting for second time and then panic in rust. the caller would have to explicitly free the buffer after use.

not sure i wanna do it in this PR though

Reply:
Yeah, this is out of scope here.

We were previously talking about introducing PaintCtx that would be passed down the component tree the same way as EventCtx is now, I wonder if that could be used for tracking the buffer state?

Reply:
made a separate issue: https://github.com/trezor/trezor-firmware/issues/2648